compiler/next/include/chpl/AST/CallExpr.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_AST_CALLEXPR_H
+#define CHPL_AST_CALLEXPR_H
+
+#include "chpl/AST/Expr.h"
+
+#include <vector>
+
+namespace chpl {
+namespace ast {
+
+/**
+  This abstract class represents something call-like
+ */
+class CallExpr : public Expr {
+ public:
+  ~CallExpr() override = 0;
+
+  // note: the reason for the +/- 1 below is that the
+  // 0'th child is the called expression, which does
+  // not count as an "actual".
+
+  int numActuals() const {
+    return this->numChildren() - 1;
+  }
+  const Expr* actual(int i) const {
+    return this->getChild(i+1);
+  }
+  Expr* baseExpr() const {
+    return this->getChild(0);
+  }
+};
```

Comment:
I suggest for `CallExpr` not to introduce `baseExpr()`. Because it means different things in CallExpr subclasses: a primitive, an operator, the class being "new"-ed, etc.

Also, factoring out `actual(i)` into `CallExpr` is not so obvious to me. Because I expect that they start at various offsets into BaseAST::children_ . For operator calls they can start at 0, whereas for new-expressions they should perhaps start at 2, after the expression for the class and the expression for the memory manager. Although maybe we can gloss over that and represent the memory manager with just a UniqueString_. Dunno how this trick will play out in the long run.

// NewExpr.h is missing from this PR

What we DO want to factor out is named actual arguments for the paren-call and new-expr subclasses. I would be OK putting support for those into CallExpr and leaving it unused for operator and primitive calls.

Reply:
I think it'd be reasonable to always use a baseExpr. For operators it could be the Identifier for the operator. For primitives it could still be... an Identifier. Even something like a `try` decorator could just use an Identifier for "try".

I'm not set on that anyway and this PR is just getting started (and in particular these call classes are currently unused and a sketch). So I don't consider any of this holding up this PR specifically.

Reply:
> I'm not set on that anyway and this PR is just getting started (and in particular these call classes are currently unused and a sketch).

In that case I'd like to remove `baseExpr` altogether and introduce it later if it passes a proper review. At which point, I'd like to see how it is used before introducing it. I expect that each subclass of `Call` will have its own variant, with a specific name and return type. The corresponding operation on a generic `Call` will be simply `child(0)`.

For numActuals() and actuals() I'd like to have a comment that they are unused for `Try` and `Dot` subclasses.

Reply:
P.S. I am OK making these changes in your follow-up PR.